util/memory/meminfo.go

```diff
@@ -200,3 +185,232 @@ func InstanceMemUsed() (uint64, error) {
 	serverMemUsage.set(memoryUsage, time.Now())
 	return memoryUsage, nil
 }
+
+// MemTotalCGroup returns the total amount of RAM that TIDB is limited to use in container environment.
+func MemTotalCGroup() (uint64, error) {
+	mem, t := memLimit.get()
+	if time.Since(t) < 60*time.Second {
+		return mem, nil
+	}
+	mem, err := getMemoryLimit()
+	if err != nil {
+		return mem, err
+	}
+	if mem == maximumBytes {
+		mem = hostTotalMem
+	}
+	memLimit.set(mem, time.Now())
+	return mem, nil
+}
+
+// MemUsedCGroup returns the total used amount of RAM on this system in container environment.
+func MemUsedCGroup() (uint64, error) {
+	mem, t := memUsage.get()
+	if time.Since(t) < 500*time.Millisecond {
+		return mem, nil
+	}
+	mem, err := getMemoryUsed()
+	if err != nil {
+		return mem, err
+	}
+	memUsage.set(mem, time.Now())
+	return mem, nil
+}
+
+// getMemoryLimit reads from memory subsystem and returns TIDB's memory limit
+func getMemoryLimit() (uint64, error) {
+	if isCgroupV1 {
+		limitInBytes, err := getCgroupFileValue(memCtrName, "memory.limit_in_bytes")
+		if err != nil {
+			return maximumBytes, err
+		}
+		if limitInBytes == uint64(maximumBytes) {
+			if isHierarchical {
+				return grepCgroupFile(memCtrName, "memory.stat", "hierarchical_memory_limit")
+			}
+		}
+		return limitInBytes, nil
+	}
+	// for cgroup v2, we get memory limit directly from /sys/fs/cgroup/memory.max
+	v, err := getCgroupFileValue("", "memory.max")
+	if err != nil {
+		return maximumBytes, nil
+	}
+	return v, nil
+}
+
+// getMemoryUsed calculate memory usage of the container
+// On cgroup v1 host, the result is `memory.usage_in_bytes - mem.Stats["total_inactive_file"]` .
+// On cgroup v2 host, the result is `memory.usage_in_bytes - mem.Stats["inactive_file"] `.
+func getMemoryUsed() (uint64, error) {
+	var (
+		ctr, memUsedFile, metric string
+		mem                      uint64
+		err                      error
+	)
+
+	if isCgroupV1 {
+		ctr, memUsedFile, metric = memCtrName, "memory.usage_in_bytes", "total_inactive_file"
+	} else {
+		ctr, memUsedFile, metric = "", "memory.current", "inactive_file"
+	}
+	mem, err = getCgroupFileValue(ctr, memUsedFile)
+	if err != nil {
+		return mem, err
+	}
+
+	v, err := grepCgroupFile(memCtrName, "memory.stat", metric)
+	if err != nil {
+		return mem, err
+	}
+	if v < mem {
+		return mem - v, nil
+	}
+	return mem, nil
+}
+
+// getHierarchicalStatus returns true when hierarchical accounting is enabled
+// find more information on https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt
+// cgroup v1 only
+func getHierarchicalStatus() bool {
+	value, _ := readCgroupFile(memCtrName, "memory.use_hierarchy")
+	return strings.TrimSpace(string(value)) == "1"
+}
+
+// isCgrpV1 tells whether we are using cgroupV1
+// see https://docs.docker.com/config/containers/runmetrics/#enumerate-cgroups
+func isCgrpV1() (bool, error) {
```

Comment:
here can also refer to runc, how to judge whether  in cgroupv2 https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/utils.go#L34